packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart

```diff
@@ -139,32 +141,91 @@ class CloudWatchLoggerPlugin extends AWSLoggerPlugin
     Future<void> startSyncing() async {
       final batchStream = _getLogBatchesToSync();
       await for (final (logs, events) in batchStream) {
-        final response = await _sendToCloudWatch(events);
-        // TODO(nikahsn): handle tooOldLogEventEndIndex
-        // and expiredLogEventEndIndex.
-        if (response.rejectedLogEventsInfo?.tooNewLogEventStartIndex != null) {
-          // TODO(nikahsn): throw and exception to enable log rotation if the
-          // log store is full.
-          break;
+        _TooNewLogEventException? tooNewException;
+        while (logs.isNotEmpty && events.isNotEmpty) {
+          final rejectedLogEventsInfo =
+              (await _sendToCloudWatch(events)).rejectedLogEventsInfo;
+          if (rejectedLogEventsInfo == null) {
+            await _logStore.deleteItems(logs);
+            break;
+          }
+          {
+            final (tooOldEndIndex, tooNewStartIndex) =
+                rejectedLogEventsInfo.parse(events.length);
+
+            if (_isValidIndex(tooNewStartIndex, events.length)) {
+              tooNewException = _TooNewLogEventException(
+                events[tooNewStartIndex!].timestamp.toInt(),
+              );
+              // set logs to end before the index.
+              logs.removeRange(tooNewStartIndex, events.length);
+              // set events to end before the index.
+              events.removeRange(tooNewStartIndex, events.length);
+            }
+            if (_isValidIndex(tooOldEndIndex, events.length)) {
+              // remove old logs from log store.
+              await _logStore.deleteItems(logs.sublist(0, tooOldEndIndex! + 1));
+              // set logs to start after the index.
+              logs.removeRange(0, tooOldEndIndex + 1);
+              // set events to start after the index.
+              events.removeRange(0, tooOldEndIndex + 1);
+            }
+            continue;
+          }
+        }
+        // after sending each batch to CloudWatch check if the batch has
+        // `tooNewException` and throw to stop syncing next batches.
+        if (tooNewException != null) {
+          throw tooNewException;
```

Comment:
I think this would be cleaner if you just returned the retry time instead of throwing